packages/federation/src/supergraph.ts

```diff
@@ -81,6 +92,51 @@ export function getFieldMergerFromSupergraphSdl(
   const memoizedASTPrint = memoize1(print);
   const memoizedTypePrint = memoize1((type: GraphQLOutputType) => type.toString());
   return function (candidates: MergeFieldConfigCandidate[]) {
+    if (
+      candidates.length === 1 ||
+      candidates.some(candidate => candidate.fieldName === '_entities')
+    ) {
+      return candidates[0].fieldConfig;
+    }
+    if (candidates.some(candidate => rootTypeMap.has(candidate.type.name))) {
+      return {
+        ...defaultMerger(candidates),
+        resolve(_root, _args, context, info) {
+          let currentSubschema: SubschemaConfig | undefined;
+          let currentScore = Infinity;
+          for (const fieldNode of info.fieldNodes) {
+            const candidatesReversed = candidates.toReversed
+              ? candidates.toReversed()
+              : [...candidates].reverse();
+            for (const candidate of candidatesReversed) {
+              const typeFieldMap = candidate.type.getFields();
+              if (candidate.transformedSubschema) {
+                const unavailableFields = extractUnavailableFields(
+                  candidate.transformedSubschema.transformedSchema,
+                  typeFieldMap[candidate.fieldName],
+                  fieldNode,
+                  () => true,
+                );
+                const score = calculateSelectionScore(unavailableFields);
+                if (score < currentScore) {
+                  currentScore = score;
+                  currentSubschema = candidate.transformedSubschema;
+                }
+              }
+            }
+          }
+          if (!currentSubschema) {
+            throw new Error('Could not determine subschema');
+          }
+          return delegateToSchema({
+            schema: currentSubschema,
+            operation: (rootTypeMap.get(info.parentType.name) || 'query') as OperationTypeNode,
```

Comment:
Instead of casting, fix the map types in the `rootTypeMap`.
```ts
const rootTypeMap = new Map<string, OperationTypeNode>
```

Reply:
I fixed the type of the Map but I have to keep `as OperationTypeNode` because it is a union of strings in GraphQLJS v15 while it is an enum in the newer GraphQL js versions :/

Reply:
Ahh. 😓 

Good enough!